### PR TITLE
firedocstore: unordered action lists

### PR DESCRIPTION
**Pull request by jba**
Implement unordered action lists in a simple way, using the same grouping
as for ordered lists. The difference is that we run the groups concurrently.

TODO: add conformance tests.

### CODE REVIEW

File: internal/docstore/firedocstore/fs.go [hunk 1]
```diff
@@ -487,6 +506,38 @@ func (c *collection) updateWrites(doc driver.Document, docName string, mods []dr
...
+			if g[0].Kind == driver.Get {
+				errs := c.runGets(ctx, g)
+				for i, err := range errs {
+					errs[base+i] = err
+				}
+			} else {
+				err := c.runWrites(ctx, g)
+				// Writes run in a transaction, so there is a single error for the group.
+				for i := 0; i < len(g); i++ {
+					errs[base+i] = err
```
Comment (shantuo) [inline internal/docstore/firedocstore/fs.go]: Why do we need to repeat the same errors and return them? It doesn't seem to help the user this way, instead, we could just return all different ones.

File: internal/docstore/firedocstore/fs.go [hunk 2]
```diff
@@ -487,6 +506,38 @@ func (c *collection) updateWrites(doc driver.Document, docName string, mods []dr
 	return []*pb.Write{w}, nil
 }
 
+func (c *collection) runActionsUnordered(ctx context.Context, actions []*driver.Action) driver.ActionListError {
+	// Split into groups the same way, but run them concurrently.
+	// TODO(jba): group without considering order.
+	groups := driver.SplitActions(actions, shouldSplit)
+	errs := make([]error, len(actions))
```
Comment (shantuo) [inline internal/docstore/firedocstore/fs.go]: `errs` doesn't need to be the length of the actions, see comment below.

File: internal/docstore/firedocstore/fs.go [hunk 3]
```diff
@@ -255,55 +259,70 @@ func shouldSplit(cur, new *driver.Action) bool {
...
+	errs := c.runGets(ctx, gets)
+	for i, err := range errs {
+		if err != nil {
+			return i, err
+		}
+	}
+	return len(gets), nil
+}
+
+// runGets executs a group of Get actions by calling the BatchGetDocuments RPC.
```
[the remark this answers is not in the document]
Comment (jba) [reply internal/docstore/firedocstore/fs.go]: Done.

File: internal/docstore/firedocstore/fs.go (hunk 1 above)
Comment (jba) [reply internal/docstore/firedocstore/fs.go]: Unordered lists have to return an error for each failed action. In this case, all the actions in the write failed. We have to a return an error for all of them.

How would we make the errors different? The only thing that's different is the action itself, but the index lets the user look up the action, so we are already encoding that information.

File: internal/docstore/firedocstore/fs.go (hunk 2 above)
Comment (jba) [reply internal/docstore/firedocstore/fs.go]: This is just a convenience. Only the non-nil errors are returned, see `driver.driver.NewActionListError`.

**Comment by jba** [comment]
PTAL.

**Pull request merged**